sway/config.c

```diff
@@ -338,34 +338,62 @@ static bool file_exists(const char *path) {
 	return path && access(path, R_OK) != -1;
 }
 
-static char *get_config_path(void) {
-	static const char *config_paths[] = {
-		"$HOME/.sway/config",
-		"$XDG_CONFIG_HOME/sway/config",
-		"$HOME/.i3/config",
-		"$XDG_CONFIG_HOME/i3/config",
-		SYSCONFDIR "/sway/config",
-		SYSCONFDIR "/i3/config",
-	};
+static char *config_path(const char *prefix, const char *config_folder) {
+	if (!prefix || !prefix[0] || !config_folder || !config_folder[0]) {
+		return NULL;
+	}
+
+	const char *filename = "config";
+
+	size_t size = 3 + strlen(prefix) + strlen(config_folder) + strlen(filename);
+	char *path = calloc(size, sizeof(char));
+	snprintf(path, size, "%s/%s/%s", prefix, config_folder, filename);
+	return path;
+}
 
-	char *config_home = getenv("XDG_CONFIG_HOME");
-	if (!config_home || !*config_home) {
-		config_paths[1] = "$HOME/.config/sway/config";
-		config_paths[3] = "$HOME/.config/i3/config";
+static char *get_config_path(void) {
+	const char *home = getenv("HOME");
+	size_t size_fallback = 1 + strlen(home) + strlen("/.config");
+	char *config_home_fallback = calloc(size_fallback, sizeof(char));
+	snprintf(config_home_fallback, size_fallback, "%s/.config", home);
+
+	const char *config_home = getenv("XDG_CONFIG_HOME");
+	if (config_home == NULL || config_home[0] == '\0') {
+		config_home = config_home_fallback;
 	}
 
-	for (size_t i = 0; i < sizeof(config_paths) / sizeof(char *); ++i) {
-		wordexp_t p;
-		if (wordexp(config_paths[i], &p, WRDE_UNDEF) == 0) {
-			char *path = strdup(p.we_wordv[0]);
-			wordfree(&p);
-			if (file_exists(path)) {
-				return path;
-			}
-			free(path);
+	struct config_path {
+		const char *prefix;
+		const char *config_folder;
+	};
+
+	static struct config_path config_paths[6];
```

Comment:
Hm, this is getting a little out of hand. Maybe can be simplified to:

```c
struct config_path config_paths[] = {
    { .prefix = "asdf", .config_folder = "asdf" },
    …
};
```

Reply:
I would like to do that, but that does only work for constant strings and not for variables.

Reply:
Put everything that is constant into the struct and added the variable bits afterwards. I think that's the best we can do, except having one array for all paths and one for all folders and a strange function which maps the order of the config files to the place in each array. But that will probably not make things clearer.

Also change it to calculate the number of tried config files at runtime.

Reply:
Maybe try dropping the `const` qualifier for the array

Reply:
You mean the static qualifier? That solved the issue. Didn't know static has some form of constant global behaviour.